api/src/main/java/org/pmiops/workbench/api/ProfileController.java

```diff
@@ -223,26 +230,57 @@ private User initializeUserIfNeeded() {
-    // initialization period for Firecloud projects, so it is not possible to do this immediately
-    // after project creation (on first sign-in).
-    if (BillingProjectStatus.PENDING.equals(user.getFreeTierBillingProjectStatus())) {
-      // If the user has not been granted the BQ job user Google role on the billing project yet,
-      // give it to them (so they can run BQ queries from notebooks).
-      try {
-        fireCloudService.grantGoogleRoleToUser(user.getFreeTierBillingProjectName(),
-            FireCloudService.BIGQUERY_JOB_USER_GOOGLE_ROLE, user.getEmail());
-      } catch (ApiException e) {
-        log.log(Level.INFO, String.format(
-            "granting BigQuery role on free tier billing project failed, this is expected " +
-            "shortly after creation of the project; will retry on next initialization attempt", e));
-        // Allow the user to continue, as most workbench functionality will still be usable.
+    // Free tier billing project setup is complete; nothing to do.
+    if (!BillingProjectStatus.PENDING.equals(user.getFreeTierBillingProjectStatus())) {
+      return user;
+    }
+
+    // On subsequent sign-ins to the first, attempt to complete the setup of the FC billing project
+    // and mark the Workbench's project setup as completed. FC project creation is asynchronous, so
+    // first confirm whether Firecloud claims the project setup is complete.
+    BillingProjectStatus status = null;
+    try {
+      status = fireCloudService.getBillingProjectMemberships().stream()
+          .filter(m -> user.getFreeTierBillingProjectName().equals(m.getProjectName()))
+          .map(m -> fcToWorkbenchBillingMap.get(m.getCreationStatus()))
+          // Should be at most one matching billing project; though we're not asserting this.
+          .findFirst()
+          .orElse(BillingProjectStatus.NONE);
+    } catch (ApiException e) {
+      log.log(Level.WARNING, "failed to retrieve billing projects, continuing", e);
+      return user;
+    }
+    switch (status) {
+      case NONE:
+      case PENDING:
+        log.log(Level.INFO, "free tier project is still initializing, continuing");
         return user;
-      }
-      user.setFreeTierBillingProjectStatus(BillingProjectStatus.READY);
-      return userDao.save(user);
+
+      case ERROR:
+        log.log(Level.SEVERE, String.format(
+            "free tier project %s failed to be created", user.getFreeTierBillingProjectName()));
+        user.setFreeTierBillingProjectStatus(status);
+        return userDao.save(user);
+
+      case READY:
+        break;
+    }
+
+    // Grant the user BQ job access on the billing project so that they can run BQ queries from
+    // notebooks.
+    try {
+      fireCloudService.grantGoogleRoleToUser(user.getFreeTierBillingProjectName(),
+          FireCloudService.BIGQUERY_JOB_USER_GOOGLE_ROLE, user.getEmail());
+    } catch (ApiException e) {
+      log.log(Level.INFO,
```

Comment:
This is no longer expected, right? (You checked status above...)

Seems like we should log severe if this happens, if not return a 500.

Reply:
Yes, forgot to update - updated and made it a warning. I don't want to return a 500 - the approach we've taken so far is that the user should be able to interact with the application, despite their free tier billing project potentially being in a bad state, or still initializing. So a transient error shouldn't block the user anymore than FC's normal delay in project initialization.